src/tracks/ui/TextEditHelper.cpp

```diff
@@ -0,0 +1,666 @@
+/*!********************************************************************
+*
+ Audacity: A Digital Audio Editor
+
+ TextEditHelper.h
+
+ Vitaly Sverchinsky
+
+ The major part of the logic is extracted from LabelTrackView.cpp and
+ LabelTextHandle.cpp
+
+ **********************************************************************/
+
+#include "TextEditHelper.h"
+
+#include <wx/app.h>
+#include <wx/dc.h>
+#include <wx/dcmemory.h>
+#include <wx/clipbrd.h>
+
+#include "../../ProjectWindow.h"
+#include "../../RefreshCode.h"
+
+#include "TextEditDelegate.h"
+
+bool TextEditHelper::IsGoodEditKeyCode(int keyCode)
+{
+    // Accept everything outside of WXK_START through WXK_COMMAND, plus the keys
+   // within that range that are usually printable, plus the ones we use for
+   // keyboard navigation.
+    return keyCode < WXK_START ||
+        (keyCode >= WXK_END && keyCode < WXK_UP) ||
+        (keyCode == WXK_RIGHT) ||
+        (keyCode >= WXK_NUMPAD0 && keyCode <= WXK_DIVIDE) ||
+        (keyCode >= WXK_NUMPAD_SPACE && keyCode <= WXK_NUMPAD_ENTER) ||
+        (keyCode >= WXK_NUMPAD_HOME && keyCode <= WXK_NUMPAD_END) ||
+        (keyCode >= WXK_NUMPAD_DELETE && keyCode <= WXK_NUMPAD_DIVIDE) ||
+#if defined(__WXMAC__)
+        (keyCode > WXK_RAW_CONTROL) ||
+#endif
+        (keyCode > WXK_WINDOWS_MENU);
+}
+
+TextEditHelper::TextEditHelper(const std::weak_ptr<TextEditDelegate>& delegate, const wxString& text, const wxFont& font)
+    : mFont(font), 
+    mDelegate(delegate), 
+    mText(text)
+{
+    mCurrentCursorPos = mInitialCursorPos = text.Length();
+}
+
+void TextEditHelper::SetTextColor(const wxColor& textColor)
+{
+    mTextColor = textColor;
+}
+
+void TextEditHelper::SetTextSelectionColor(const wxColor& textSelectionColor)
+{
+    mTextSelectionColor = textSelectionColor;
+}
+
+void TextEditHelper::Cancel(AudacityProject* project)
+{
+    if (auto lock = mDelegate.lock())
+        lock->OnTextEditCancelled(project);
+}
+
+void TextEditHelper::Finish(AudacityProject* project)
+{
+    if (auto lock = mDelegate.lock())
+        lock->OnTextEditFinished(project, mText);
+}
+
+std::pair<int, int> TextEditHelper::GetSelection() const
+{
+    return std::make_pair(mInitialCursorPos, mCurrentCursorPos);
+}
+
+void TextEditHelper::SetSelection(int from, int to)
+{
+    mInitialCursorPos = from;
+    mCurrentCursorPos = to;
```

Comment:
Should you comment before this function,
```
@pre from <= to
```

Or swap the values before the assignment to make the condition hold?